modules/apps/callbacks/ibc_middleware.go

```diff
@@ -0,0 +1,243 @@
+		panic("underlying application does not implement PacketDataUnmarshaler")
+	}
+	return IBCMiddleware{
+		app:            packetUnmarshalerApp,
+		channel:        channel,
+		contractKeeper: contractKeeper,
+	}
+}
+
+// UnmarshalPacketData defers to the underlying app to unmarshal the packet data.
+func (im IBCMiddleware) UnmarshalPacketData(bz []byte) (interface{}, error) {
+	return im.app.UnmarshalPacketData(bz)
+}
+
+// OnAcknowledgementPacket implements source callbacks for acknowledgement packets.
+// It defers to the underlying application and then calls the contract callback.
+// If the contract callback fails (within the gas limit), state changes are reverted.
+func (im IBCMiddleware) OnAcknowledgementPacket(
+	ctx sdk.Context,
+	packet channeltypes.Packet,
+	acknowledgement []byte,
+	relayer sdk.AccAddress,
+) error {
+	// we first call the underlying app to handle the acknowledgement
+	appResult := im.app.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)
+	if appResult != nil {
+		return appResult
+	}
+
+	callbackData, err := types.GetCallbackData(im.app, packet, ctx.GasMeter().GasRemaining())
+	if err != nil {
+		types.EmitSourceCallbackEvent(ctx, packet, types.CallbackTypeAcknowledgement, callbackData, err)
+		return appResult
+	}
+	if callbackData.SrcContractAddr == "" {
+		return appResult
+	}
+
+	var ack channeltypes.Acknowledgement
+	if err := channeltypes.SubModuleCdc.UnmarshalJSON(acknowledgement, &ack); err != nil {
+		err = errorsmod.Wrapf(ibcerrors.ErrUnknownRequest, "cannot unmarshal callback packet acknowledgement: %v", err)
+		types.EmitSourceCallbackEvent(ctx, packet, types.CallbackTypeAcknowledgement, callbackData, err)
+		return appResult
+	}
+
+	cachedCtx, writeFn := ctx.CacheContext()
+	cachedCtx = cachedCtx.WithGasMeter(sdk.NewGasMeter(callbackData.GasLimit))
+
+	err = im.contractKeeper.IBCAcknowledgementPacketCallback(cachedCtx, packet, ack, relayer, callbackData.SrcContractAddr)
+	if err == nil {
+		writeFn()
+	}
+	// consume gas
+	ctx.GasMeter().ConsumeGas(cachedCtx.GasMeter().GasConsumed(), "ibc acknowledgement packet callback")
+
+	// emit event as a callback success
+	types.EmitSourceCallbackEvent(ctx, packet, types.CallbackTypeAcknowledgement, callbackData, err)
+	return appResult
+}
+
+// OnTimeoutPacket implements timeout source callbacks for the ibc-callbacks middleware.
+// It defers to the underlying application and then calls the contract callback.
+// If the contract callback fails (within the gas limit), state changes are reverted.
+func (im IBCMiddleware) OnTimeoutPacket(ctx sdk.Context, packet channeltypes.Packet, relayer sdk.AccAddress) error {
+	appResult := im.app.OnTimeoutPacket(ctx, packet, relayer)
+	if appResult != nil {
+		return appResult
+	}
+
+	callbackData, err := types.GetCallbackData(im.app, packet, ctx.GasMeter().GasRemaining())
+	if err != nil {
+		types.EmitSourceCallbackEvent(ctx, packet, types.CallbackTypeTimeoutPacket, callbackData, err)
+		return appResult
```

Comment:
ditto here. If appResult is nil then just return nil. Also document why

Reply:
done. I've done a big refactor of these functions.